README
Term-ReadLine::Simple
===================


INFO

Read a line.



INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install



SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Term::ReadLine::Simple



LICENSE AND COPYRIGHT

Copyright (C) 2014-2015 Matthäus Kiem

This library is free software; you
can redistribute it and/or modify it under the same terms
as Perl 5.10.0. For details, see the full text of the
licenses in the file LICENSE.


Changes
Revision history for Term::ReadLine::Simple


0.308   2015-12-13
        - Handle multi-line prompt.

0.307   2015-12-04
        - Makefile.PL: fixed url to github repository  [github #1 - Mohammad S Anwar (Manwar)].

0.306   2015-12-03
        - List::Util: 'any()'not available before version "1.33".

0.305   2015-12-02
        - Bugfix in "fill_form": don't alter the passed AoA - make a copy instead.
        - Removed option "sep".
        - New option "ro" - readonly.
        - Update documentation.

0.304   2015-11-12
        - Bugfix list-reference.

0.303   2015-10-31
        - Bugfix: set "auto_up" to 0 so it is defined.

0.302   2015-05-11
        - "readline": catch "key-up" and "key-down".

0.301   2015-04-14
        - Expect tests only in xt.
        - Code refactoring.

0.300   2015-03-28
        - "fill_form": meaning of the option "auto_up" changed.
        - "fill_form": new option "sep".

0.209   2015-03-23
        - Changed tests.

0.208_06   2015-03-22
        - Changed tests.

0.208_05   2015-03-21
        - Changed tests.

0.208_04   2015-03-20
        - Changed tests.

0.208_03   2015-03-19
        - Changed tests.

0.208_02   2015-03-19
        - Changed tests.

0.208_01   2015-03-18
        - Changed tests.

0.208   2015-03-18
        - Changed tests.

0.207_02   2015-03-18
        - Changed tests.

0.207_01   2015-03-17
        - Changed tests.

0.207   2015-03-17
        - option "compat" affects also "fill_form".
        - Update documentation.

0.206   2015-03-16
        - Modified test.

0.205   2015-03-16
        - Modified test.
        - Changed "ENTER" key behavior.

0.204   2015-03-16
        - Modified test.

0.203   2015-03-15
        - Modified test.

0.202   2015-03-15
        - Refactored "option" code.

0.201   2015-03-15
        - "fill_form": new option "auto_up".
        - Code refactoring.
        - Update documentation.

0.200   2015-03-14
        - Refactored 'readline' - keep output in one row.
        - New method 'fill_form'.
        - Update documentation.

0.016   2015-02-05
        - Use constant.
        - Update year copyright.

0.015   2014-12-08
        - Perl minimum version: "5.8.3".
        - Update build requirements.

0.014   2014-11-20
        - Make tests which use "Expect" optional.

0.013   2014-09-06
        - Modified tests: skip if a "Given filehandle is not a tty in clone_winsize_from".

0.012   2014-08-27
        - Improved documentation.

0.011   2014-08-27
        - First release on CPAN.


t/readline.pl
#!/usr/bin/env perl
use warnings;
use strict;
use 5.008003;

use Term::ReadLine::Simple;

use FindBin qw( $RealBin );
use lib $RealBin;
use Data_Test_Readline;

my $a_ref = Data_Test_Readline::return_test_data();
my $args  = $a_ref->[shift]{arguments};

my $tiny = Term::ReadLine::Simple->new();
my $line = $tiny->readline( @$args );

print "<$line>\n";


t/Data_Test_Readline.pm
package # hide
Data_Test_Readline;

use 5.008003;
use warnings;
use strict;


sub key_seq {
    return {
        CONTROL_M => "\x{0d}",
        ENTER     => "\x{0d}",

        RIGHT     => "\e[C",
        RIGHT_O   => "\eOC",
        CONTROL_F => "\x{06}",

        LEFT      => "\e[D",
        LEFT_O    => "\eOD",
        CONTROL_B => "\x{02}",

        CONTROL_A => "\x{01}",
        HOME      => "\e[H",

        CONTROL_E => "\x{05}",
        END       => "\e[F",

        CONTROL_H => "\x{08}",
        BTAB      => "\x{08}",
        BSPACE    => "\x{7f}",
        BTAB_Z    => "\e[Z",
        BTAB_OZ   => "\eOZ",

        CONTROL_D => "\x{04}",
        DELETE    => "\e[3~",

        CONTROL_K => "\x{0b}",
        CONTROL_U => "\x{15}",
    };
}

sub return_test_data {
    return [
        {
            used_keys => [ "-", 'ENTER' ],
            expected  => "<->",
            arguments => [ 'Prompt: ' ],
        },
        {
            used_keys => [ 'ENTER' ],
            expected  => "<default>",
            arguments => [ 'Prompt: ', "default" ],
        },
        {
            used_keys => [ ( 'LEFT' ) x 4, 'CONTROL_K', " ", 'ENTER' ],
            expected  => "<def >",
            arguments => [ 'Prompt: ', 'default' ],
        },
        {
            used_keys => [ ( 'LEFT' ) x 5, "-", 'ENTER' ],
            expected  => "<de-fault>",
            arguments => [ 'Prompt: ', "default" ],
        },
        {
            used_keys => [ ( 'BSPACE' ) x 7, "hello", " ", "world", 'ENTER' ],
            expected  => "<hello world>",
            arguments => [ 'Prompt: ', { default => "default" } ],
        },
        {
            used_keys => [ "house", 'ENTER' ],
            expected  => "<The house>",
            arguments => [ 'Prompt: ', { default => "The ", no_echo => 1 } ],
        },
        {
            used_keys => [ "0123456789", ( 'LEFT' ) x 5, 'CONTROL_U', 'ENTER' ],
            expected  => "<56789>",
            arguments => [ ': ' ],
        },
        {
            used_keys => [ "The black cat climbed the green tree", 'HOME', ( 'RIGHT' ) x 4, ( 'DELETE' ) x 6, 'END', ( 'LEFT' ) x 5, ( 'BSPACE' ) x 6, 'ENTER' ],
            expected  => "<The cat climbed the tree>",
            arguments => [ 'Prompt: ' ],
        },
        {
            used_keys => [ "abcde ", "XY " x 10, ( 'LEFT' ) x 36, ( 'RIGHT' ) x 8, 'CONTROL_K', 'ENTER' ],
            expected  => "<DEFAULT abcde XY>",
            arguments => [ "Prompt: ", { default => "DEFAULT " } ],
        },
        {
            used_keys => [ qw( HOME RIGHT RIGHT_O DELETE CONTROL_D END LEFT LEFT_O BSPACE CONTROL_H CONTROL_A DELETE CONTROL_E BSPACE _ABC ENTER ) ],
            expected  => "<hblack cat climbed the green e_ABC>",
            arguments => [ 'Prompt: ', { default => "The black cat climbed the green tree" } ],
        },
        {
             used_keys => [ 'CONTROL_U', "a", 'ENTER'],
             expected  => "<a>",
             arguments => [ 'Prompt: ', { default => "many words " x 100, no_echo => 0 } ],
        },
    ];
}




1;

__END__


t/Data_Test_Arguments.pm
package # hide
Data_Test_Arguments;

use 5.008003;
use warnings;
use strict;



sub valid_args {
    return [
        {
            expected => "<>",
            args => [ 'Prompt: ' ],
        },
        {
            expected => "<default>",
            args => [ 'Prompt: ', 'default' ],
        },
        {
            expected => "<default>",
            args => [ 'Prompt: ', { default => 'default', no_echo => 0 } ],
        },
        {
            expected => "<DEFAULT>",
            args => [ 'Prompt: ', { no_echo => 1, default => "DEFAULT" } ],
        },
        {
            expected => "<123>",
            args => [ "Prompt: ", { no_echo => 2, default => '123' } ],
        },
        {
            expected => "<123>",
            args => [ "Prompt: ", { no_echo => 2, default => 123 } ],
        },
    ];
}


sub invalid_args {
    return [
        {
            expected => "<readline: a reference is not a valid prompt.>",
            args => [ {} ],
        },
        {
            expected => "<readline: a reference is not a valid prompt.>",
            args => [ [] ],
        },
        {
            expected => "<readline: the (optional) second argument must be a string or a HASH reference>",
            args => [ 'Prompt: ', [] ],
        },
        {
            expected => "<readline: option 'default' : a reference is not a valid value.>",
            args => [ 'Prompt: ', { default => [] } ],
        },
        {
            expected => "<readline: option 'default' : a reference is not a valid value.>",
            args => [ 'Prompt: ', { default => {} } ],
        },
        {
            expected => "<readline: option 'no_echo' : '-1' is not a valid value.>",
            args => [ 'Prompt: ', { no_echo => -1 } ],
        },
        {
            expected => "<readline: option 'no_echo' : '3' is not a valid value.>",
            args => [ 'Prompt: ', { no_echo => 3 } ],
        },
        {
            expected => "<readline: option 'no_echo' : 'a' is not a valid value.>",
            args => [ 'Prompt: ', { no_echo => 'a' } ],
        },
        {
            expected => "<readline: option 'no_echo' : a reference is not a valid value.>",
            args => [ 'Prompt: ', { no_echo => [] } ],
        },
        {
            expected => "<readline: option 'no_echo' : a reference is not a valid value.>",
            args => [ 'Prompt: ', { no_echo => {} } ],
        },
        {
            expected => "<readline: option 'default' : a reference is not a valid value.>",
            args => [ 'Prompt: ', { default => {}, no_echo => 1 } ],
        },
        {
            expected => "<readline: 'hello' is not a valid option name>",
            args => [ 'Prompt: ', { default => 'default', hello => 1, no_echo => 0 } ],
        },
    ];
}


1;

__END__


lib/Term/ReadLine/Simple/Linux.pm
package # hide from PAUSE
Term::ReadLine::Simple::Linux;

use warnings;
use strict;
use 5.008003;

our $VERSION = '0.308';

use Term::ReadKey  qw( GetTerminalSize ReadKey ReadMode );

use Term::ReadLine::Simple::Constants qw( :linux );


sub new {
    return bless {}, $_[0];
}


sub __set_mode {
    #my ( $self ) = @_;
    ReadMode( 'cbreak' );
};


sub __reset_mode {
    #my ( $self ) = @_;
    ReadMode( 'restore' );
}


sub __term_buff_size {
    #my ( $self ) = @_;
    my ( $term_width, $term_height ) = GetTerminalSize();
    return $term_width, $term_height;
}


sub __get_key {
    #my ( $self ) = @_;
    my $c1 = ReadKey( 0 );
    return if ! defined $c1;
    if ( $c1 eq "\e" ) {
        my $c2 = ReadKey( 0.10 );
        if ( ! defined $c2 ) {
            return  NEXT_get_key; # KEY_ESC
        }
        elsif ( $c2 eq 'O' ) {
            my $c3 = ReadKey( 0 );
               if ( $c3 eq 'A' ) { return VK_UP; }
            elsif ( $c3 eq 'B' ) { return VK_DOWN; }
            elsif ( $c3 eq 'C' ) { return VK_RIGHT; }
            elsif ( $c3 eq 'D' ) { return VK_LEFT; }
            elsif ( $c3 eq 'F' ) { return VK_END; }
            elsif ( $c3 eq 'H' ) { return VK_HOME; }
            elsif ( $c3 eq 'Z' ) { return KEY_BTAB; }
            else {
                return NEXT_get_key;
            }
        }
        elsif ( $c2 eq '[' ) {
            my $c3 = ReadKey( 0 );
               if ( $c3 eq 'A' ) { return VK_UP; }
            elsif ( $c3 eq 'B' ) { return VK_DOWN; }
            elsif ( $c3 eq 'C' ) { return VK_RIGHT; }
            elsif ( $c3 eq 'D' ) { return VK_LEFT; }
            elsif ( $c3 eq 'F' ) { return VK_END; }
            elsif ( $c3 eq 'H' ) { return VK_HOME; }
            elsif ( $c3 eq 'Z' ) { return KEY_BTAB; }
            elsif ( $c3 =~ /^[0-9]$/ ) {
                my $c4 = ReadKey( 0 );
                if ( $c4 eq '~' ) {
                       if ( $c3 eq '3' ) { return VK_DELETE; }
                    elsif ( $c3 eq '5' ) { return VK_PAGE_UP; }
                    elsif ( $c3 eq '6' ) { return VK_PAGE_DOWN; }
                    else {
                        return NEXT_get_key;
                    }
                }
                else {
                    return NEXT_get_key;
                }
            }
            else {
                return NEXT_get_key;
            }
        }
        else {
            return NEXT_get_key;
        }
    }
    else {
        return ord $c1;
    }
};


sub __up    {
    return if ! $_[1];
    print "\e[${_[1]}A";
}

sub __down  {
    return if ! $_[1];
    print "\e[${_[1]}B";
}

sub __left  {
    return if ! $_[1];
    print "\e[${_[1]}D"; }

sub __right {
    return if ! $_[1];
    print "\e[${_[1]}C";
}

sub __reverse { print "\e[7m"; }

sub __reset   { print "\e[0m"; }

sub __mark_current { print "\e[4m"; } # "\e[1m\e[4m";

sub __clear_screen { print "\e[H\e[J"; }

sub __clear_lines_to_end_of_screen { print "\r\e[0J"; }

sub __clear_line { print "\r\e[K"; }

sub __beep {
#    print "\a";
}

1;

__END__


lib/Term/ReadLine/Simple/Win32.pm
package # hide from PAUSE
Term::ReadLine::Simple::Win32;

use warnings;
use strict;
use 5.008003;

our $VERSION = '0.308';

use Encode qw( decode );

use Encode::Locale qw();
use Win32::Console qw( STD_INPUT_HANDLE ENABLE_PROCESSED_INPUT STD_OUTPUT_HANDLE
                       RIGHT_ALT_PRESSED LEFT_ALT_PRESSED RIGHT_CTRL_PRESSED LEFT_CTRL_PRESSED SHIFT_PRESSED
                       FOREGROUND_INTENSITY BACKGROUND_INTENSITY );

use Term::ReadLine::Simple::Constants qw( :win32 );


sub new {
    return bless {}, $_[0];
}


sub __set_mode {
    my ( $self ) = @_;
    $self->{input} = Win32::Console->new( STD_INPUT_HANDLE );
    $self->{old_in_mode} = $self->{input}->Mode();
    $self->{input}->Mode( ENABLE_PROCESSED_INPUT );
    $self->{output} = Win32::Console->new( STD_OUTPUT_HANDLE );
    $self->{def_attr}  = $self->{output}->Attr();
    $self->{bg_color}  = $self->{def_attr} & 0x70;
    $self->{fill_attr} = $self->{bg_color} | $self->{bg_color};
}


sub __reset_mode {
    my ( $self ) = @_;
    if ( defined $self->{input} ) {
        if ( defined $self->{old_in_mode} ) {
            $self->{input}->Mode( $self->{old_in_mode} );
            delete $self->{old_in_mode};
        }
        $self->{input}->Flush;
        # workaround Bug #33513:
        delete $self->{input}{handle};
    }
    if ( defined $self->{output} ) {
        delete $self->{output}{handle}; # ?
    }
}


sub SHIFTED_MASK () {
      RIGHT_ALT_PRESSED
    | LEFT_ALT_PRESSED
    | RIGHT_CTRL_PRESSED
    | LEFT_CTRL_PRESSED
    | SHIFT_PRESSED
}

sub __get_key {
    my ( $self ) = @_;
    my @event = $self->{input}->Input;
    my $event_type = shift @event;
    return NEXT_get_key if ! defined $event_type;
    if ( $event_type == 1 ) {
        my ( $key_down, $repeat_count, $v_key_code, $v_scan_code, $char, $ctrl_key_state ) = @event;
        return NEXT_get_key if ! $key_down;
        if ( $char ) {
            return ord decode( 'console_in', chr( $char & 0xff ) );
        }
        else{
            if ( $ctrl_key_state & SHIFTED_MASK ) {
                return NEXT_get_key;
            }
            elsif ( $v_key_code == VK_CODE_END )       { return VK_END }
            elsif ( $v_key_code == VK_CODE_HOME )      { return VK_HOME }
            elsif ( $v_key_code == VK_CODE_LEFT )      { return VK_LEFT }
            elsif ( $v_key_code == VK_CODE_UP )        { return VK_UP }
            elsif ( $v_key_code == VK_CODE_DOWN )      { return VK_DOWN }
            elsif ( $v_key_code == VK_CODE_RIGHT )     { return VK_RIGHT }
            elsif ( $v_key_code == VK_CODE_PAGE_UP )   { return VK_PAGE_UP }
            elsif ( $v_key_code == VK_CODE_PAGE_DOWN ) { return VK_PAGE_DOWN }
            elsif ( $v_key_code == VK_CODE_DELETE )    { return VK_DELETE }
            else {
                return NEXT_get_key;
            }
        }
    }
    else {
        return NEXT_get_key;
    }
}


sub __term_buff_size {
    my ( $self ) = @_;
    my ( $term_width, $term_height ) = $self->{output}->MaxWindow();
    return $term_width - 1, $term_height;
}

sub __get_cursor_position {
    my ( $self ) = @_;
    my ( $col, $row ) = $self->{output}->Cursor();
    return $col, $row;
}

sub __set_cursor_position {
    my ( $self, $col, $row ) = @_;
    $self->{output}->Cursor( $col, $row );
}

sub __up {
    my ( $self, $rows_up ) = @_;
    return if ! $rows_up;
    my ( $col, $row ) = $self->__get_cursor_position;
    $self->__set_cursor_position( $col, $row - $rows_up  );
}

sub __down {
    my ( $self, $rows_down ) = @_;
    return if ! $rows_down;
    my ( $col, $row ) = $self->__get_cursor_position;
    $self->__set_cursor_position( $col, $row + $rows_down  );
}

sub __left {
    my ( $self, $cols_left ) = @_;
    return if ! $cols_left;
    my ( $col, $row ) = $self->__get_cursor_position;
    $self->__set_cursor_position( $col - $cols_left, $row  );
}

sub __right {
    my ( $self, $cols_right ) = @_;
    return if ! $cols_right;
    my ( $col, $row ) = $self->__get_cursor_position;
    $self->__set_cursor_position( $col + $cols_right, $row  );
}

sub __clear_screen {
    my ( $self ) = @_;
    $self->{output}->Cls( $self->{def_attr} );
}

sub __clear_lines_to_end_of_screen {
    my ( $self ) = @_;
    my ( $width, $height ) = $self->{output}->Size(); #
    my ( $col, $row ) = $self->__get_cursor_position();
    $self->__set_cursor_position( 0, $row  );
    $self->{output}->FillAttr(
            $self->{fill_attr},
            $width * $height, #
            0, $row );
}

sub __clear_line {
    my ( $self ) = @_;
    my ( $width, $height ) = $self->{output}->Size(); #
    my ( $col, $row ) = $self->__get_cursor_position();
    $self->__set_cursor_position( 0, $row  );
    $self->{output}->FillAttr(
            $self->{fill_attr},
            $width,
            0, $row );
}

sub __reverse {
    my ( $self ) = @_;
    $self->{output}->Attr( $self->{inverse} );
}

sub __reset {
    my ( $self ) = @_;
    $self->{output}->Attr( $self->{def_attr} );
}

sub __mark_current {
    my ( $self ) = @_;
    $self->{output}->Attr( $self->{def_attr} | FOREGROUND_INTENSITY | BACKGROUND_INTENSITY  );
}

sub __beep {}



1;

__END__


lib/Term/ReadLine/Simple/Constants.pm
package # hide from PAUSE
Term::ReadLine::Simple::Constants;

use warnings;
use strict;
use 5.008003;

our $VERSION = '0.308';

use Exporter qw( import );

our @EXPORT_OK = qw(
        NEXT_get_key
        CONTROL_A CONTROL_B CONTROL_D CONTROL_E CONTROL_F CONTROL_H CONTROL_K CONTROL_U KEY_BTAB KEY_TAB
        KEY_ENTER KEY_ESC KEY_BSPACE
        VK_CODE_END VK_CODE_HOME VK_CODE_LEFT VK_CODE_UP VK_CODE_RIGHT VK_CODE_DOWN VK_CODE_DELETE VK_CODE_PAGE_UP VK_CODE_PAGE_DOWN
        VK_END      VK_HOME      VK_LEFT      VK_UP      VK_RIGHT      VK_DOWN      VK_DELETE      VK_PAGE_UP      VK_PAGE_DOWN
);

our %EXPORT_TAGS = (
    rl => [ qw(
        NEXT_get_key
        CONTROL_A CONTROL_B CONTROL_D CONTROL_E CONTROL_F CONTROL_H CONTROL_K CONTROL_U KEY_BTAB KEY_TAB
        KEY_ENTER KEY_ESC KEY_BSPACE
        VK_END VK_HOME VK_LEFT VK_UP VK_RIGHT VK_DOWN VK_DELETE VK_PAGE_UP VK_PAGE_DOWN
    ) ],
    linux  => [ qw(
        NEXT_get_key
        KEY_BTAB KEY_ESC
        VK_END VK_HOME VK_LEFT VK_UP VK_RIGHT VK_DOWN VK_DELETE VK_PAGE_UP VK_PAGE_DOWN
    ) ],
    win32  => [ qw(
        NEXT_get_key
        VK_CODE_END VK_CODE_HOME VK_CODE_LEFT VK_CODE_UP VK_CODE_RIGHT VK_CODE_DOWN VK_CODE_DELETE VK_CODE_PAGE_UP VK_CODE_PAGE_DOWN
        VK_END      VK_HOME      VK_LEFT      VK_UP      VK_RIGHT      VK_DOWN      VK_DELETE      VK_PAGE_UP      VK_PAGE_DOWN
    ) ]
);


use constant {
    NEXT_get_key => -1,

    CONTROL_A  => 0x01,
    CONTROL_B  => 0x02,
    CONTROL_D  => 0x04,
    CONTROL_E  => 0x05,
    CONTROL_F  => 0x06,
    CONTROL_H  => 0x08,
    KEY_BTAB   => 0x08,
    KEY_TAB    => 0x09,
    CONTROL_K  => 0x0b,
    KEY_ENTER  => 0x0d,
    CONTROL_U  => 0x15,
    KEY_ESC    => 0x1b,
    KEY_BSPACE => 0x7f,

    VK_PAGE_UP   => 333,
    VK_PAGE_DOWN => 334,
    VK_END       => 335,
    VK_HOME      => 336,
    VK_LEFT      => 337,
    VK_UP        => 338,
    VK_RIGHT     => 339,
    VK_DOWN      => 340,
    VK_DELETE    => 346,

    VK_CODE_PAGE_UP   => 33,
    VK_CODE_PAGE_DOWN => 34,
    VK_CODE_END       => 35,
    VK_CODE_HOME      => 36,
    VK_CODE_LEFT      => 37,
    VK_CODE_UP        => 38,
    VK_CODE_RIGHT     => 39,
    VK_CODE_DOWN      => 40,
    VK_CODE_DELETE    => 46,
};



1;

__END__


Makefile.PL
use 5.008003;
use warnings;
use strict;
use ExtUtils::MakeMaker;


my %prereqs_os_specific;
my %test_prereqs_os_specific;
if ( $^O eq 'MSWin32' ) {
    %prereqs_os_specific = (
        'Win32::Console' => 0,
    );
}
else {
    %prereqs_os_specific = (
        'Term::ReadKey'  => 0,
    );
}


WriteMakefile(
    PL_FILES         => {},
    MIN_PERL_VERSION => '5.008003',
    LICENSE          => 'perl',
    META_MERGE => {
        'meta-spec' => { version => 2 },
        resources   => {
            repository => {
              type => 'git',
              web  => 'https://github.com/kuerbis/Term-ReadLine-Simple',
              url  => 'https://github.com/kuerbis/Term-ReadLine-Simple.git',
            },
        },
    },
    AUTHOR           => 'Matthaeus Kiem <cuer2s@gmail.com>',
    NAME             => 'Term::ReadLine::Simple',
    ABSTRACT_FROM    => 'lib/Term/ReadLine/Simple.pm',
    VERSION_FROM     => 'lib/Term/ReadLine/Simple.pm',
    EXE_FILES        => [],
    BUILD_REQUIRES   => {
        'File::Spec::Functions' => 0,
        'FindBin'               => 0,
        'Test::More'            => 0,
        'Test::Fatal'           => 0,
    },
    PREREQ_PM => {
        'strict'            => 0,
        'warnings'          => 0,
        'Carp'              => 0,
        'Encode'            => 0,
        'Encode::Locale'    => 0,
        'Exporter'          => 0,
        'List::Util'        => '1.33',
        'Unicode::GCString' => 0,
        %prereqs_os_specific,
    },
    dist  => {
        COMPRESS => 'gzip',
        SUFFIX   => '.gz',
        TARFLAGS => '--format=gnu -cvf',
    },
    clean => {
        FILES => 'Term-ReadLine-Simple-*',
    },
);
